File: frontend/src/types/events.ts
export type EventType =
  | 'session.created'
  | 'session.ended'
  | 'transcript.partial'
  | 'transcript.final'
  | 'intent.detected'
  | 'tool.started'
  | 'tool.completed'
  | 'rag.retrieved'
  | 'policy.decision'
  | 'workflow.step'
  | 'response.generated'
  | 'sentiment.updated'
  | 'escalation.created'
  | 'call.summary'
  | 'audio.completed'
  | 'customer.updated'
  | 'invoice.updated'
  | 'appointment.updated'
  | 'error'

export interface BaseEvent {
  event: EventType
  session_id: string
  timestamp: string
}

export interface SessionCreatedEvent extends BaseEvent {
  event: 'session.created'
  conversation_id: string
  customer_id: string | null
  channel: string
}

export interface SessionEndedEvent extends BaseEvent {
  event: 'session.ended'
  duration_sec: number | null
}

export interface TranscriptPartialEvent extends BaseEvent {
  event: 'transcript.partial'
  text: string
}

export interface TranscriptFinalEvent extends BaseEvent {
  event: 'transcript.final'
  text: string
  turn_index: number | null
}

export interface IntentDetectedEvent extends BaseEvent {
  event: 'intent.detected'
  intents: string[]
  entities: Record<string, string>
  sentiment: string
  urgency: string
  confidence: number | null
}

export interface ToolStartedEvent extends BaseEvent {
  event: 'tool.started'
  tool_name: string
  input_params: Record<string, unknown>
}

export interface ToolCompletedEvent extends BaseEvent {
  event: 'tool.completed'
  tool_name: string
  status: string
  input_params?: Record<string, unknown>
  output: Record<string, unknown>
  duration_ms: number
}

export interface RagRetrievedEvent extends BaseEvent {
  event: 'rag.retrieved'
  query: string
  passages: Array<{ text: string; doc_id: string; score: number }>
  doc_count: number
}

export interface PolicyDecisionEvent extends BaseEvent {
  event: 'policy.decision'
  policy_name: string
  action_proposed: string
  authorized: boolean
  reason: string
}

export interface WorkflowStepEvent extends BaseEvent {
  event: 'workflow.step'
  workflow_name: string
  step_name: string
  step_status: string
  steps_completed: string[]
}

export interface ResponseGeneratedEvent extends BaseEvent {
  event: 'response.generated'
  text: string
}

export interface SentimentUpdatedEvent extends BaseEvent {
  event: 'sentiment.updated'
  sentiment: string
  urgency: string | null
}

export interface ErrorEvent extends BaseEvent {
  event: 'error'
  code: string
  message: string
}

export interface EscalationCreatedEvent extends BaseEvent {
  event: 'escalation.created'
  reason: string
  domain: string
  sentiment: string
  turn_count: number
  customer_verified: boolean
}

export interface CallSummaryEvent extends BaseEvent {
  event: 'call.summary'
  summary_text: string
  resolution: string
  escalated: boolean
  duration_sec: number
  tools_used: string[]
}

export interface CustomerUpdatedEvent extends BaseEvent {
  event: 'customer.updated'
  customer_id: string
}

export interface InvoiceUpdatedEvent extends BaseEvent {
  event: 'invoice.updated'
  customer_id: string
  invoice_id: string
}

export interface AppointmentUpdatedEvent extends BaseEvent {
  event: 'appointment.updated'
  customer_id: string
  appointment_id: string
}

export type AnyEvent =
  | SessionCreatedEvent
  | SessionEndedEvent
  | TranscriptPartialEvent
  | TranscriptFinalEvent
  | IntentDetectedEvent
  | ToolStartedEvent
  | ToolCompletedEvent
  | RagRetrievedEvent
  | PolicyDecisionEvent
  | WorkflowStepEvent
  | ResponseGeneratedEvent
  | SentimentUpdatedEvent
  | EscalationCreatedEvent
  | CallSummaryEvent
  | CustomerUpdatedEvent
  | InvoiceUpdatedEvent
  | AppointmentUpdatedEvent
  | ErrorEvent

File: backend/app/api/websocket/events.py
from enum import Enum
from typing import Any
from pydantic import BaseModel
from datetime import datetime, timezone


class EventType(str, Enum):
    SESSION_CREATED = "session.created"
    SESSION_ENDED = "session.ended"
    TRANSCRIPT_PARTIAL = "transcript.partial"
    TRANSCRIPT_FINAL = "transcript.final"
    INTENT_DETECTED = "intent.detected"
    TOOL_STARTED = "tool.started"
    TOOL_COMPLETED = "tool.completed"
    RAG_RETRIEVED = "rag.retrieved"
    POLICY_DECISION = "policy.decision"
    WORKFLOW_STEP = "workflow.step"
    RESPONSE_GENERATED = "response.generated"
    SENTIMENT_UPDATED = "sentiment.updated"
    AUDIO_STARTED = "audio.started"
    AUDIO_COMPLETED = "audio.completed"
    ESCALATION_CREATED = "escalation.created"
    CALL_SUMMARY = "call.summary"
    ERROR = "error"
    CUSTOMER_UPDATED = "customer.updated"
    INVOICE_UPDATED = "invoice.updated"
    APPOINTMENT_UPDATED = "appointment.updated"


class BaseEvent(BaseModel):
    event: EventType
    session_id: str
    timestamp: str = ""

    def model_post_init(self, __context: Any):
        if not self.timestamp:
            self.timestamp = datetime.now(timezone.utc).isoformat()


class SessionCreatedEvent(BaseEvent):
    event: EventType = EventType.SESSION_CREATED
    conversation_id: str
    customer_id: str | None = None
    channel: str = "web"


class SessionEndedEvent(BaseEvent):
    event: EventType = EventType.SESSION_ENDED
    duration_sec: int | None = None


class TranscriptPartialEvent(BaseEvent):
    event: EventType = EventType.TRANSCRIPT_PARTIAL
    text: str


class TranscriptFinalEvent(BaseEvent):
    event: EventType = EventType.TRANSCRIPT_FINAL
    text: str
    turn_index: int | None = None


class IntentDetectedEvent(BaseEvent):
    event: EventType = EventType.INTENT_DETECTED
    intents: list[str]
    entities: dict
    sentiment: str
    urgency: str
    confidence: float | None = None


class ToolStartedEvent(BaseEvent):
    event: EventType = EventType.TOOL_STARTED
    tool_name: str
    input_params: dict


class ToolCompletedEvent(BaseEvent):
    event: EventType = EventType.TOOL_COMPLETED
    tool_name: str
    status: str
    output: dict
    duration_ms: int


class RagRetrievedEvent(BaseEvent):
    event: EventType = EventType.RAG_RETRIEVED
    query: str
    passages: list[dict]
    doc_count: int


class PolicyDecisionEvent(BaseEvent):
    event: EventType = EventType.POLICY_DECISION
    policy_name: str
    action_proposed: str
    authorized: bool
    reason: str


class WorkflowStepEvent(BaseEvent):
    event: EventType = EventType.WORKFLOW_STEP
    workflow_name: str
    step_name: str
    step_status: str
    steps_completed: list[str]


class ResponseGeneratedEvent(BaseEvent):
    event: EventType = EventType.RESPONSE_GENERATED
    text: str


class SentimentUpdatedEvent(BaseEvent):
    event: EventType = EventType.SENTIMENT_UPDATED
    sentiment: str
    urgency: str | None = None


class ErrorEvent(BaseEvent):
    event: EventType = EventType.ERROR
    code: str
    message: str


class EscalationCreatedEvent(BaseEvent):
    event: EventType = EventType.ESCALATION_CREATED
    reason: str
    domain: str
    sentiment: str
    turn_count: int
    customer_verified: bool


class CallSummaryEvent(BaseEvent):
    event: EventType = EventType.CALL_SUMMARY
    summary_text: str
    resolution: str
    escalated: bool
    duration_sec: int
    tools_used: list[str]


class CustomerUpdatedEvent(BaseEvent):
    event: EventType = EventType.CUSTOMER_UPDATED
    customer_id: str


class InvoiceUpdatedEvent(BaseEvent):
    event: EventType = EventType.INVOICE_UPDATED
    customer_id: str
    invoice_id: str


class AppointmentUpdatedEvent(BaseEvent):
    event: EventType = EventType.APPOINTMENT_UPDATED
    customer_id: str
    appointment_id: str
